rfc-100-linting.md

```diff
@@ -0,0 +1,55 @@
+
+## Issues
+
+1. Rubocop now has a native way of sharing rules: you include rules by specifing `inherit_gem` in `.rubocop.yml`. This has been successfully [trialed in content-performance-manager](https://github.com/alphagov/content-performance-manager/pull/1082). As mentioned in that PR, this makes the linting faster and makes it compatible with developer environments (see [this issue from 2016](https://github.com/alphagov/govuk-lint/issues/61) for evidence of a need).
+1. A lot of hard work has gone into fixing older Ruby linting violations ([for example in finder-frontend](https://github.com/alphagov/finder-frontend/pull/579)). This has allowed us to [turn off the "diffing" behaviour](https://github.com/alphagov/finder-frontend/pull/581) for most projects - we now lint all of the code all of the time.
+1. We currently do not lint our Javascript. This is a gap first raised [in 2016](https://github.com/alphagov/govuk-lint/issues/51) and further discussed in an [issue talking about Standard.js](https://github.com/alphagov/govuk-lint/issues/63).
+1. The [scss_lint tool is being deprecated](https://github.com/alphagov/govuk-lint/issues/70). The authors suggest using a different library.
+1. The CSS linting hasn't been fully adopted in GOV.UK projects - only a few frontend applications have it enabled. Others [explicitly disable it](https://github.com/search?q=org%3Aalphagov+sassLint%3A+false&type=Code). The main reason for this is that `scss_lint` does not have an autocorrect feature, so it's a significant (and boring) investment to adopt the linting for older projects.
+1. While govuk-lint has been adopted by [a lot of Ruby projects in GDS and wider government](https://github.com/alphagov/govuk-lint/network/dependents), the CSS linting feature isn't always used. Projects that don't use CSS still need to pull in the `scss_lint` dependency.
+1. Last year, [GOV.UK Frontend](https://github.com/alphagov/govuk-frontend) (part of the [Design System](https://design-system.service.gov.uk/)) was officially launched. Since this is the defacto standard for building frontend things in government, GOV.UK should be adopting the same tools it uses. This will allow us to easily push things upstream and re-use GOV.UK Frontend patterns.
+1. There's been a shift in the Ruby community regarding the use of Javascript packages. Where previously it was preferred to use tools written in Ruby like `scss_lint`, Rails now [ships with Yarn](https://guides.rubyonrails.org/5_1_release_notes.html#yarn-support) and supports [Webpack](https://guides.rubyonrails.org/5_1_release_notes.html#optional-webpack-support).
+
+## Proposal
+
+1. Deprecate the `govuk-lint` gem
+2. Create a new gem called `govuk_rubocop` that includes all rules configuration and a dependency on the `rubocop` gem
+3. Adopt NPM modules [standard](https://www.npmjs.com/package/standard) for Javascript linting and [sass-lint](https://www.npmjs.com/package/sass-lint) for CSS linting
```

Comment:
The open question here is whether or not we need to share config for sass-lint. GOV.UK Frontend [uses a YAML file to configure it](https://github.com/alphagov/govuk-frontend/blob/master/config/.sass-lint.yml). The preference would be to have these rules defined in a single place. @alex-ju is going to investigate this.

Reply:
At one point we were definitely talking about shipping our linter config either as part of GOV.UK Frontend or as a separate package – if this would be useful it _might_ be something the team can look into. (/cc @nickcolley @hannalaakso)

Reply:
Yeah, providing they're not too big we could ship them as part of GOV.UK Frontend main package which would be relatively easy.

Alternatively we could consider publishing a separate package, which may be OK since they dont get updated very often, and easier to contribute to since it'd be a small repository.

Reply:
@36degrees that would be really helpful as we won't need to create a separate repo/place to keep these sort of config files.

Reply:
Are the rules in GOV.UK frontend sass-lint consistent with the ones used in the scss_lint tool?

If not do we know whether we're going to switch to GOV.UK Frontend rules or maintain the scss_lint rules?

Reply:
@kevindew not the same rules they are different and GOV.UK frontend sass-lint are more constraining; `*.scss` need to be [updated](https://github.com/alphagov/govuk_publishing_components/pull/758/commits/2959153de1604cb995a2b11d8c7c66363a91cd4d) in each project before moving to `sass-lint`. Good thing is that they both work in parallel for a while - no clashing rules.